src/resources/dac_esp32.cc

```diff
@@ -0,0 +1,236 @@
+  return dac_channel_t(-1);
+}
+
+#elif CONFIG_IDF_TARGET_ESP32
+
+#error "Unsupported ESP32 target"
+
+#else
+
+static dac_channel_t get_dac_channel(int pin) {
+  return dac_channel_t(-1);
+}
+
+#endif
+
+class DacResourceGroup;
+
+// TODO(florian): access to this variable needs to be protected by a lock.
+static int _cosine_user_count = 0;
+
+class DacResource : public Resource {
+ public:
+  TAG(DacResource);
+  DacResource(DacResourceGroup* group, dac_channel_t channel);
+
+  virtual ~DacResource() override {
+    unuse_cosine();
+  }
+
+  dac_channel_t channel() const { return _channel; }
+
+  esp_err_t use_cosine() {
+    esp_err_t err = ESP_OK;
+    if (_uses_cosine) return err;
+    _uses_cosine = true;
+    // TODO(florian): this needs to be protected in a lock.
+    _cosine_user_count++;
+    if (_cosine_user_count == 1) {
+      // First user.
+      err = dac_cw_generator_enable();
+    }
+    return err;
+  }
+
+  esp_err_t unuse_cosine() {
+    esp_err_t err = ESP_OK;
+    if (!_uses_cosine) return err;
+    _uses_cosine = false;
+    // TODO(florian): this needs to be protected in a lock.
+    _cosine_user_count--;
+    if (_cosine_user_count == 0) {
+      // Last user.
+      err = dac_cw_generator_disable();
+    }
+    return err;
+  }
+
+ private:
+  dac_channel_t _channel;
+  bool _uses_cosine = false;
+};
+
+class DacResourceGroup : public ResourceGroup {
+ public:
+  TAG(DacResourceGroup);
+  explicit DacResourceGroup(Process* process)
+      : ResourceGroup(process) {}
+
+  virtual void on_unregister_resource(Resource* resource) override {
+    DacResource* dac_resource = static_cast<DacResource*>(resource);
+
+    dac_output_disable(dac_resource->channel());
+  }
+ private:
+  static int _cosine_user_count;
+};
+
+DacResource::DacResource(DacResourceGroup* group, dac_channel_t channel)
+    : Resource(group), _channel(channel) {}
+
+
+MODULE_IMPLEMENTATION(dac, MODULE_DAC)
+
+PRIMITIVE(init) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  DacResourceGroup* touch = _new DacResourceGroup(process);
+  if (!touch) MALLOC_FAILED;
+
+  proxy->set_external_address(touch);
+  return proxy;
+}
+
+PRIMITIVE(use) {
+  ARGS(DacResourceGroup, group, int, pin, uint8, initial_value);
+
+  dac_channel_t channel = get_dac_channel(pin);
+  // TODO(florian): The compiler complains that the result of the comparison
+  // is always false (`-Wtautological-constant-out-of-range-compare`). I believe we use
+  // this in many other places. Do we care?
+  if (channel == -1) INVALID_ARGUMENT;
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  // TODO(florian): the AdcState is allocated with a `HeapTagScope`. Do we need
+  // to do the same here?
+  DacResource* resource = _new DacResource(group, channel);
+  if (resource == null) MALLOC_FAILED;
+  // TODO(florian): if we fail, do we need to release the resource that isn't
```

Comment:
Yes, you need to delete the resource before returning prior to registering the resource.

Reply:
done.